internal/db/schema/migrations/oss/postgres/22/01_session_valid_state.up.sql

```diff
@@ -0,0 +1,93 @@
+begin;
+
+-- Remove invalid (states with no prior end time) rows
+delete from session_state where previous_end_time is null and state != 'pending';
+
+-- In the case of dupes that are otherwise valid, identify first valid session
+create temp table session_state_delete
+as select session_id, min(start_time) as st_t from session_state ou
+   where (select count(*) from session_state inr
+          where inr.session_id = ou.session_id and inr.state = ou.state) > 1
+   group by session_id;
+
+-- Remove all session rows after first valid session
+DELETE FROM session_state t1
+    using session_state_delete t2 where t1.session_id =t2.session_id and t1.start_time > t2.st_t;
+
+drop table session_state_delete;
```

Comment:
I think this can be done without creating a temp table, either by using a CTE or a sub-query for the delete. Since this is a one-off query that happens in a migration, the performance considerations between those three approaches is probably negligible, but I would have a preference towards the CTE or sub-query so this step is all in one sql statement.